src/ucp/tag/offload/rndv.c

```diff
@@ -0,0 +1,232 @@
+        ucp_send_request_id_alloc(req);
+    }
+
+    return status;
+}
+
+static ucs_status_t
+ucp_tag_rndv_offload_send_func(ucp_request_t *req,
+                               const ucp_proto_single_priv_t *spriv,
+                               uct_iov_t *iov)
+{
+    ucp_tag_offload_unexp_rndv_hdr_t rndv_hdr = {
+        .ep_id    = ucp_send_request_get_ep_remote_id(req),
+        .req_id   = ucp_send_request_get_id(req),
+        .md_index = spriv->super.md_index
+    };
+    void *rndv_op;
+
+    rndv_op = uct_ep_tag_rndv_zcopy(ucp_ep_get_fast_lane(req->send.ep,
+                                                         spriv->super.lane),
+                                    req->send.msg_proto.tag,
+                                    &rndv_hdr, sizeof(rndv_hdr), iov, 1, 0,
+                                    &req->send.state.uct_comp);
+
+    if (ucs_unlikely(UCS_PTR_IS_ERR(rndv_op))) {
+        return UCS_PTR_STATUS(rndv_op);
+    }
+
+    req->flags                   |= UCP_REQUEST_FLAG_OFFLOADED;
+    req->send.tag_offload.rndv_op = rndv_op;
+
+    return UCS_OK;
+}
+
+UCS_PROFILE_FUNC(ucs_status_t, ucp_tag_rndv_offload_proto_progress, (self),
+                 uct_pending_req_t *self)
+{
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
+
+    return ucp_proto_zcopy_single_progress(
+            req, UCT_MD_MEM_ACCESS_RMA | UCT_MD_MEM_FLAG_HIDE_ERRORS,
+            ucp_tag_rndv_offload_send_func, NULL,
+            ucp_tag_rndv_offload_completion, ucp_tag_rndv_offload_request_init);
+}
+
+ucp_proto_t ucp_tag_rndv_offload_proto = {
+    .name     = "tag/rndv/offload",
+    .desc     = "rendezvous tag offload",
+    .flags    = 0,
+    .init     = ucp_tag_rndv_offload_proto_init,
+    .query    = ucp_proto_single_query,
+    .progress = {ucp_tag_rndv_offload_proto_progress},
+    .abort    = ucp_proto_request_zcopy_abort,
+    .reset    = ucp_proto_request_zcopy_id_reset
+};
+
+/* SW emulation of rndv offload protocol. This protocol sends regular RTS as
+ * offloaded eager message and this RTS is always handled by SW on the receiver.
+ * It can be used when real rndv offload is not supported (e.g. for non-contig
+ * or very large messages) or when it is enabled by protocol selection.
+ */
+static ucs_status_t
+ucp_tag_rndv_offload_sw_proto_init(const ucp_proto_init_params_t *init_params)
+{
+    ucp_worker_h worker        = init_params->worker;
+    ucp_context_h context      = worker->context;
+    ucp_ep_config_t *ep_config = &ucs_array_elem(&worker->ep_config,
+                                                 init_params->ep_cfg_index);
+    ucp_proto_rndv_ctrl_init_params_t params = {
+        .super.super         = *init_params,
+        .super.latency       = 0,
+        .super.overhead      = 40e-9,
+        .super.cfg_thresh    = context->config.ext.rndv_thresh,
+        .super.cfg_priority  = 60,
+        .super.min_length    = ucp_ep_tag_offload_min_rndv_thresh(ep_config),
```

Comment:
let's try to not reuse protov1 logic

Reply:
this restriction is still relevant

Reply:
why? I can try to invoke `ucp_rkey_packed_size` directly, but this would be the same as we do when calculating `config->rndv.rkey_size

Reply:
when we remove protov1 we would want to remove most of config contents.

Reply:
updated `ucp_ep_tag_offload_min_rndv_thresh`